universal-application-tool-0.0.1/app/services/LocalizedStrings.java

```diff
@@ -0,0 +1,262 @@
+  public boolean hasTranslationFor(Locale locale) {
+    return translations().containsKey(locale);
+  }
+
+  /** Returns the set of locales these localized strings support. */
+  public ImmutableSet<Locale> locales() {
+    return translations().keySet();
+  }
+
+  /** Returns true if these localized strings have no translations. */
+  @JsonIgnore
+  public boolean isEmpty() {
+    return translations().isEmpty();
+  }
+
+  /** Returns true if any of the translations are blank strings. */
+  public boolean hasEmptyTranslation() {
+    return translations().values().stream().anyMatch(String::isBlank);
+  }
+
+  /**
+   * Attempts to get the translation for the given locale, and falls back on the translation for the
+   * default locale if it is missing.
+   *
+   * @throws RuntimeException if it tries to use the translation for the default locale and it is
+   *     missing.
+   */
+  public String getOrDefault(Locale locale) {
+    try {
+      return get(locale);
+    } catch (TranslationNotFoundException e) {
+      return getDefault();
+    }
+  }
+
+  /**
+   * Gets the translation for CiviForm's default locale.
+   *
+   * @throws RuntimeException if the translation for the default locale is missing.
+   */
+  @JsonIgnore
+  public String getDefault() {
+    try {
+      return get(DEFAULT_LOCALE, true);
+    } catch (TranslationNotFoundException e) {
+      // This should never happen - US English should always be supported.
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * Get an {@link Optional} containing the translation for the locale, or {@link Optional#empty()}
+   * if the locale is not supported.
+   */
+  public Optional<String> maybeGet(Locale locale) {
+    try {
+      return Optional.of(get(locale, true));
+    } catch (TranslationNotFoundException e) {
+      return Optional.empty();
+    }
+  }
+
+  /**
+   * Get the translation for the given locale.
+   *
+   * <p>If the localized strings are not required (see {@link #isRequired()}, and there are no
+   * translations, this will return the empty string.
+   *
+   * @throws TranslationNotFoundException if the locale is not supported.
+   */
+  public String get(Locale locale) throws TranslationNotFoundException {
+    return get(locale, isRequired());
+  }
+
+  /**
+   * Get the translation for the given locale.
+   *
+   * <p>If these localized strings are not required (see {@link #isRequired()}), and there are no
+   * translations, then return the empty string.
+   *
+   * @param isRequired if true, allows the empty string to be thrown if there are no locales
+   *     supported. If there is at least one locale supported, then it still throws {@link
+   *     TranslationNotFoundException} if the locale is not supported.
+   * @throws TranslationNotFoundException if translations are required, but the locale is not
+   *     supported.
+   */
+  private String get(Locale locale, boolean isRequired) throws TranslationNotFoundException {
```

Comment:
I'll add more comments surrounding the `required` property.

I was thinking that the property should only affect `get(Locale)` and `getOrDefault(Locale)` - they return empty strings if there are no translations.

For `maybeGet(Locale)` I was thinking it makes sense to return `Optional.empty()`, whether it is required or not. If `required` affected `maybeGet(Locale)` then it would return `Optional.of("")` if there were no translations. 

What do you think is clearer? I think they're both a bit confusing. Comments will help clear it up, but which way do you think we should go on this?

Reply:
FYI the original `maybeGet` for question help text had a check to convert an empty string to an empty optional.